frontend/lib/resolution/scope-queries.cpp

```diff
@@ -891,24 +1179,103 @@ static bool lookupInScopeViz(Context* context,
     config |= LOOKUP_DECLS;
 
     doLookupInScope(context, scope, {}, resolving, name, config,
-                    checkedScopes, improperMatches);
+                    checkedScopes, improperMatches,
+                    foundExternBlock,
+                    /* traceCurPath */ nullptr,
+                    /* traceResult */ nullptr);
   }
 
   return got;
 }
 
+static bool helpLookupInScope(Context* context,
+                              const Scope* scope,
+                              llvm::ArrayRef<const Scope*> receiverScopes,
+                              const ResolvedVisibilityScope* resolving,
+                              UniqueString name,
+                              LookupConfig config,
+                              NamedScopeSet& checkedScopes,
+                              std::vector<BorrowedIdsWithName>& result,
+                              std::vector<VisibilityTraceElt>* traceCurPath,
+                              std::vector<ResultVisibilityTrace>* traceResult)
+{
+  bool checkExternBlocks = (config & LOOKUP_EXTERN_BLOCKS) != 0;
+  bool foundExternBlock = false;
+  size_t startSize = result.size();
+  NamedScopeSet savedCheckedScopes;
+
+  if (checkExternBlocks) {
+    // clear the extern blocks lookup (since it is a 2nd pass)
+    config &= ~LOOKUP_EXTERN_BLOCKS;
+    // make a note of the checked scopes so we can reset it
+    savedCheckedScopes = checkedScopes;
+  }
+
+  doLookupInScope(context, scope, receiverScopes, resolving,
+                  name, config,
+                  checkedScopes, result,
+                  foundExternBlock,
+                  traceCurPath, traceResult);
+
+  // If we found any extern blocks, and there were no other symbols,
+  // and extern block lookup was requested, use extern block lookup.
+  if (checkExternBlocks && startSize == result.size() && foundExternBlock) {
```

Comment:
doesn't `doLookupInScope` return a `found` bool? Why not store it here / on line 1214 and then avoid the `startSize == result.size()` comparison/ even saving startSize?